### DISCUSSION TITLE
Loading all relations after object creation

### DISCUSSION THREAD

**Discussion by klaydev** [sqlalchemy/sqlalchemy#10700] [Usage Questions] [answered]
Hi. I have a task to load all relations after creating an object (`session.add(obj); session.commit()`)
Is it possible after creating an object to immediately load all its relationships by analogy with `session.get(ObjClass, obj_id, options=[joinedload('*')]`?

**Comment by klaydev** [comment]
This need is due to the fact that my model has methods that work with relations

**Comment by zzzeek** [comment]
well that get() call would be pretty close to what you want, or you can try [session.refresh()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.refresh), naming the attributes you want in the `attribute_names` parameter.

**Reply by klaydev** [reply]
But as I understand it, will it work until the session is closed? I also need to store the attributes of the object being created even after the session is closed. Correct me if I'm wrong.

**Reply by zzzeek** [reply, marked as answer]
the object stays populated as long as its not expired.   expiration occurs if you say `session.rollback()`, **or** if you leave `expire_on_commit=True` and use `session.commit()`.  If you use `session.close()` only, the objects become detached and are not expired.

**Reply by klaydev** [reply]
Yeah. You understand what my situation is. Here is my code:
```py
# database.py
class Base(DeclarativeBase):
    async def create(self):
        async with session_maker() as session:
            session.add(self)
            await session.commit()
```
```py
# models.py
class SomeModel(Base):
    some_relations = relationship(...)
    async def create(self):
        await super().create()
        await SomeDependsModel(some_model_id=self.id, some_relations_param=self.some_relations).create()
```
And i get this error:
```py
sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <SomeModel at 0x1b92046b210> is not bound to a Session; lazy load operation of attribute 'some_relations' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)
```

**Reply by zzzeek** [reply]
usually the best approach is to keep the session open for longer, letting the object go to flow within your program in a detached state is not the recommended pattern.   that object represents a row in the database so to the degree you can keep it associated with the DB, your overall operation can more closely be thought of as a whole transaction.

for lazyloading under asyncio I'd use [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs)

**Comment by klaydev** [comment]
Thank you, I will try to use AsyncAttrs, but for my purposes there is no need to keep the session open all the time, because any change to the object is done through a certain method, which opens a new session

**Reply by zzzeek** [reply]
sure, I was noting that kind of architecture is working against SQLAlchemy's main use case, though you can always take the detached objects you have and re `.add()` them to that new session, where they will again know how to load their collections and related objects.

**Comment by klaydev** [comment]
Thank you, @zzzeek , for help!
